compositor_pipeline/src/pipeline/output/whip.rs

```diff
@@ -0,0 +1,246 @@
+
+    #[error("WHIP request failed!\nMethod: {0}\nURL: {1}")]
+    RequestFailed(Method, Url),
+
+    #[error(
+        "Unable to get location endpoint, check correctness of WHIP endpoint and your Bearer token"
+    )]
+    MissingLocationHeader,
+
+    #[error("Invalid endpoint URL: {1}")]
+    InvalidEndpointUrl(#[source] ParseError, String),
+
+    #[error("Missing Host in endpoint URL")]
+    MissingHost,
+
+    #[error("Missing port in endpoint URL")]
+    MissingPort,
+
+    #[error("Failed to create RTC session description: {0}")]
+    RTCSessionDescriptionError(webrtc::Error),
+
+    #[error("Failed to set local description: {0}")]
+    LocalDescriptionError(webrtc::Error),
+
+    #[error("Failed to set remote description: {0}")]
+    RemoteDescriptionError(webrtc::Error),
+
+    #[error("Failed to parse {0} response body: {1}")]
+    BodyParsingError(&'static str, reqwest::Error),
+
+    #[error("Failed to create offer: {0}")]
+    OfferCreationError(webrtc::Error),
+
+    #[error(transparent)]
+    PeerConnectionInitError(#[from] webrtc::Error),
+
+    #[error("Failed to convert ICE candidate to JSON: {0}")]
+    IceCandidateToJsonError(webrtc::Error),
+
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::Error),
+
+    #[error(transparent)]
+    PayloadingError(#[from] PayloadingError),
+
+    #[error("Trickle ICE not supported")]
+    TrickleIceNotSupported,
+
+    #[error("Entity Tag missing")]
+    EntityTagMissing,
+
+    #[error("Entity Tag non-matching")]
+    EntityTagNonMatching,
+}
+
+const WHIP_INIT_TIMEOUT: Duration = Duration::from_secs(60);
+
+impl WhipSender {
+    pub fn new(
+        output_id: &OutputId,
+        options: WhipSenderOptions,
+        packets_receiver: Receiver<EncoderOutputEvent>,
+        request_keyframe_sender: Option<Sender<()>>,
+        pipeline_ctx: &PipelineCtx,
+    ) -> Result<Self, OutputInitError> {
+        let payloader = Payloader::new(options.video, options.audio);
+        let packet_stream = PacketStream::new(packets_receiver, payloader, 1400);
+        let should_close = Arc::new(AtomicBool::new(false));
+        let (init_confirmation_sender, mut init_confirmation_receiver) =
+            oneshot::channel::<Result<(), WhipError>>();
+
+        let whip_ctx = WhipCtx {
+            options: options.clone(),
+            output_id: output_id.clone(),
+            request_keyframe_sender,
+            should_close: should_close.clone(),
+            tokio_rt: pipeline_ctx.tokio_rt.clone(),
+        };
+
+        pipeline_ctx.tokio_rt.spawn(
+            run_whip_sender_task(
+                whip_ctx,
+                packet_stream,
+                init_confirmation_sender,
+                pipeline_ctx.event_emitter.clone(),
+            )
+            .instrument(span!(
+                Level::INFO,
+                "WHIP sender",
+                output_id = output_id.to_string()
+            )),
+        );
+
+        let start_time = Instant::now();
```

Comment:
Let's leave it for now, but this is sth that I would normally refactor into something cleaner. It is not bad on its own, but this is quite high level function, and retry logic is hiding what it is about.